Colorcode cdn

chromoxdor opened the pull request:
> > i maybe also better start to learn c++ to integrate it properly
> 
> Yep, that's the true reason I did not fix it for you. Absolutely nothing to do with me forgetting, for sure ;)

You weren’t joking 🙃 

It is less than a year ago when i wrote my first line of javascript. So i thought to myself...how hard can it be to start to understand another language even if i start with copy & paste...

But i am pretty sure that there are still a few things to be optimized.

chromoxdor committed:
test

chromoxdor committed:
Merge branch 'letscontrolit:mega' into colorcode_cdn

chromoxdor committed:
having fun :)

chromoxdor commented:
and for now the cdn links to my repository

chromoxdor committed:
order of appearance is important here

chromoxdor commented:
Question is if it is the best way to implement?

The bare latest eg. 8266_1m  mega build is 882.352 byte
The same build with plain adding of the links (like in my other pull request) is 882.784 byte
And with this approach it is 882.816 bytes

One problem of having the files locally is that the eat up ram and the ui becomes very laggy...

TD-er commented:
> You weren’t joking 🙃

Well I was, but then all my attention and time was spent on this other bug regarding ESP32 and wrong flash settings which may even loose settings on the ESP.
But glad your impatience won ;)

I will later this evening have a look at the code.

chromoxdor commented:
> Well I was, but then all my attention and time was spent on this other bug regarding ESP32 and wrong flash settings which may even loose settings on the ESP.

I know, i followed the conversation about it a bit and though that this issue is your main priority right now. That’s why i started myself.

chromoxdor committed:
Merge branch 'letscontrolit:mega' into colorcode_cdn

TD-er commented:
Hmm not sure if I should be using this PR, or #4088 
I have added something to the #4088 PR to generate CDN prefix URL on the fly during build, or even allow to override it in the Custom.h.
But perhaps that should have been a separate PR....

Anyway, which color code files are the latest?
The ones in this PR, or in #4088 or on your Git repo?

chromoxdor committed:
disabled touch to replace

chromoxdor commented:
So... i am back in business.


> Anyway, which color code files are the latest?

The files in this PR and in my repo are the latest...  
On #4088 i put the files in a separate folder since the got their own url. 
The question is, which kind of implementation you prefer?

Edit: i synced the files also with #4088

chromoxdor commented:
ubbsi...i closed it?.. now i got confused a bit.  i still struggle with github...

chromoxdor committed:
Merge branch 'colorcode_cdn' of https://github.com/chromoxdor/ESPEasy into colorcode_cdn

TD-er commented:
Should I reopen it?

chromoxdor commented:
yep...thing is, that when i sync a fork with changes of the same file i can only discard my commits.. and than i fetched this with github desktop and it asks me what to do with these conflicts. i guess i choose the wrong option.... what is the right way to deal with conflicts?

TD-er commented:
Depends on what Git tool you're using.

Here on GitHub, you can press the button at the end of the PR to "merge" the mega branch onto the branch of your PR.
That's by far the easiest way and the most secure way.
On merge conflicts you will be presented an edit window where you can choose which parts of the code you will keep.

If the merge is too complex, you must use another git tool.
I myself use GitKraken, which allows me to simply drag and drop a label to perform merge operations like these.

Never ever try to "rebase" your branch as that will for sure give you a lot of head aches.

chromoxdor commented:
Good to know...thanks!


> a lot of head aches.

i already have... :)

chromoxdor committed:
Merge branch 'mega' into colorcode_cdn

chromoxdor committed:
update webstaticdata.cpp

chromoxdor committed:
Merge branch 'letscontrolit:mega' into colorcode_cdn

chromoxdor committed:
added keywords, maybe better autocorrect handling

added P022 & P059

chromoxdor committed:
p022 ".pwm" update

chromoxdor committed:
Merge branch 'mega' into colorcode_cdn

chromoxdor committed:
Merge branch 'mega' into colorcode_cdn

chromoxdor committed:
P126 added missing  "ShiftOut,SetAllHigh"

chromoxdor commented:
@TD-er what else is missing/ has to be added/changed for merging this into mega?

TD-er commented on src/src/Static/WebStaticData.cpp:
```diff
@@ -150,4 +150,52 @@ void serve_JS(JSfiles_e JSfile) {
     streamFromFS(fname);
     html_add_script_end();
+}
+void serve_CMJS(CMfiles_e CMfile) {
```
Why a new function which does essentially the same as the one for the other JS files?

chromoxdor replied:
because i am still learning and this is my amateurish attempt to integrate the files. And i thought i might be good to have a separation to everything else.....🤷🏻‍♂️

TD-er replied:
It has 2 main drawbacks:

- Code duplication
- Increase build size

The latter one is specific an issue when working with microcontrollers.
But the first one is (IMHO) always a concern as you may also duplicate bugs or introduce bugs if you need to change something but don't change it for all occurrences.

But I think it is better to merge it and then let me or @tonhuisman deduplicate it.

chromoxdor replied:
> But I think it is better to merge it and then let me or @tonhuisman deduplicate it.

Thats a good plan. 🙂

TD-er commented on src/src/Static/WebStaticData.cpp:
```diff
@@ -150,4 +150,52 @@ void serve_JS(JSfiles_e JSfile) {
+    }
+}
+
+void serve_CMCSS() {
```
Same question here.

tonhuisman commented on static/espeasy.js:
```diff
@@ -79,7 +79,7 @@ var commonPlugins = [
   "multirelay", "multirelay,on", "multirelay,off", "multirelay,set", "multirelay,get", "multirelay,loop",
   //P126
-  "ShiftOut", "ShiftOut,Set", "ShiftOut,SetNoUpdate", "ShiftOut,Update", "ShiftOut,SetAll", "ShiftOut,SetAllNoUpdate", "ShiftOut,SetAllLow", "ShiftOut,SetChipCount", "ShiftOut,SetHexBin",
+  "ShiftOut", "ShiftOut,Set", "ShiftOut,SetNoUpdate", "ShiftOut,Update", "ShiftOut,SetAll", "ShiftOut,SetAllNoUpdate", "ShiftOut,SetAllLow", "ShiftOut,SetAllHigh", "ShiftOut,SetChipCount", "ShiftOut,SetHexBin",
```
I'd expect some of the [P129] [ShiftIn commands](https://espeasy.readthedocs.io/en/latest/Plugin/P129.html#commands) here as well?

chromoxdor replied:
And you shall have them.... 🙂

tonhuisman replied:
It's actually P129, but at least they're in now 😜 

Once this is merged developers should update this list/file when adding commands, maybe you could add that as a 'task' to the developer documentation?

chromoxdor replied:
i don’t know what happend...the 9 went upside down....

> maybe you could add that as a 'task' to the developer documentation?

where do i find this?

tonhuisman replied:
> where do i find this?

Ii think it should be part of this chapter: https://github.com/letscontrolit/ESPEasy/blob/mega/docs/source/Participate/PlatformIO.rst#writing-documentation
Or add an extra paragraph for explaining how to update these files.

chromoxdor committed:
Added  P126 "ShiftIn,.."

chromoxdor committed:
[P129]  corrected command positions

chromoxdor committed:
[P129] added main command

having the "main-command! as an extra keyword makes code-completion easier by first  choose and complete the "main-command" and then by entering "," + start to type choosing the "sub-command"

chromoxdor commented:
Since the code for implementing this looks so much better here: https://github.com/letscontrolit/ESPEasy/pull/4088😁 this can almost be closed... any objections?

TD-er commented:
> any objections?

Nope...
